pkg/ebpf/netlink/consumer.go

```diff
@@ -0,0 +1,291 @@
+package netlink
+
+import (
+	"errors"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"golang.org/x/sys/unix"
+
+	"github.com/mdlayher/netlink"
+	"github.com/mdlayher/netlink/nlenc"
+)
+
+const (
+	netlinkCtNew   = uint32(1)
+	ipctnlMsgCtGet = 1
+	outputBuffer   = 100
+)
+
+var msgBufferSize int
+
+func init() {
+	msgBufferSize = os.Getpagesize()
+}
+
+var errShortErrorMessage = errors.New("not enough data for netlink error code")
+
+// Consumer is responsible for encapsulating all the logic of hooking into Conntrack
+// consuming [NEW] connection events, and exposing a stream of the NAT connections.
+type Consumer struct {
+	conn      *netlink.Conn
+	socket    *Socket
+	pool      *bufferPool
+	workQueue chan func()
+}
+
+// Event encapsulates the result of a single netlink.Con.Receive() call
+type Event struct {
+	msgs   []netlink.Message
+	err    error
+	buffer *[]byte
+	pool   *bufferPool
+}
+
+// Messages returned from the socket read
+func (e *Event) Messages() []netlink.Message {
+	return e.msgs
+}
+
+// Error associated to the socket read
+func (e *Event) Error() error {
+	err := e.err
+	if err != nil {
+		e.Done()
+	}
+	return err
+}
+
+// Done must be called after decoding events so the underlying buffers can be reclaimed.
+func (e *Event) Done() {
+	if e.buffer != nil {
+		e.pool.Put(e.buffer)
+	}
+}
+
+func NewConsumer(procRoot string) (*Consumer, error) {
+	c := &Consumer{
+		pool:      newBufferPool(),
+		workQueue: make(chan func()),
+	}
+	c.initWorker(procRoot)
+
+	// Create netlink socket within root namespace
+	var err error
+	c.do(true, func() {
+		c.socket, err = NewSocket(c.pool)
+		if err != nil {
+			return
+		}
+
+		c.conn = netlink.NewConn(c.socket, c.socket.pid)
+
+		// Sometimes the conntrack flushes are larger than the socket recv buffer capacity.
+		// This ensures that in case of buffer overrun the `recvmsg` call will *not*
+		// receive an ENOBUF which is currently not handled properly by go-conntrack library.
+		c.conn.SetOption(netlink.NoENOBUFS, true)
```

Comment:
I know this is a POC, but let's check in the safer behavior which will stop processing if we get an enobuf and then add smarter handling